### ISSUE TITLE
Linux: Command palette disagree with shortcuts behaviour 

### ISSUE THREAD

**Issue by dayer3** [Microsoft/vscode#14984]
Hi. I don't know if it's related only to my locale and keyboard, to Linux with a keyboard layout different from US or if it's related to #713. I apologize in advance because I haven't known find it.
According to the command palette and the keybinding widget, I get some shortcuts with a different behaviour regarding I use it. This affects to shortcuts with symbols like slash, plus, minus,..

My system:

- VSCode Version: 1.6.1 and 1.7.1
- OS Version: Fedora 24 (Linux 4.7.6-200 x86_64)

```
$ locale
LANG=es_ES.UTF-8
LC_CTYPE="es_ES.UTF-8"
LC_NUMERIC="es_ES.UTF-8"
LC_TIME="es_ES.UTF-8"
LC_COLLATE="es_ES.UTF-8"
LC_MONETARY="es_ES.UTF-8"
LC_MESSAGES="es_ES.UTF-8"
LC_PAPER="es_ES.UTF-8"
LC_NAME="es_ES.UTF-8"
LC_ADDRESS="es_ES.UTF-8"
LC_TELEPHONE="es_ES.UTF-8"
LC_MEASUREMENT="es_ES.UTF-8"
LC_IDENTIFICATION="es_ES.UTF-8"
LC_ALL=
```
```
$ setxkbmap -query
rules:      evdev
model:      pc105
layout:     es,es
variant:    ,
options:    compose:rctrl
```

Keyboard like this:
![Spanish KB](https://cloud.githubusercontent.com/assets/243047/20006354/6791b778-a297-11e6-9884-f5b2552244ae.png)

Steps to Reproduce (toggle comment, for example):

1. `Ctrl` + `Shift` + `P` to look for _Toggle Line Comment_.
![Looking for command](https://cloud.githubusercontent.com/assets/243047/20005908/3b77c166-a295-11e6-8042-b0a82cb60dc2.png)
Or I look for it in the keybinding settings file:
![kb settings](https://cloud.githubusercontent.com/assets/243047/20006595/8ec589cc-a298-11e6-9767-2d11fdb184a9.png)
2. According to the suggestion, I push `Ctrl` + `-` (next to right shift key) but the font size is reduced. I don't get the toggle line comment action.
3. If I push `Ctrl` + `-` (minus sign from numpad) I get nothing.
4. If I push `Ctrl` + `ç` I get the toggle line coment.
5. I check `Ctrl` + `k` `Ctrl` + `ç` and I get this is detected like `Ctrl` + `k` `Ctrl` + `-`
![Check keybinding](https://cloud.githubusercontent.com/assets/243047/20006194/c7f135d6-a296-11e6-9e0f-b10d94654fd4.png)

I get a similar behaviour with _Toggle Integrated Terminal_ and _Split Editor_:

![integrated terminal](https://cloud.githubusercontent.com/assets/243047/20006669/eedf8b28-a298-11e6-93c7-8bfe0cc7a33d.png)
I push `Ctrl` + `º` but I get the _Split Editor_ action.

![split editor](https://cloud.githubusercontent.com/assets/243047/20006718/2b2273ca-a299-11e6-8f6a-6655f7e86b39.png)
I push `Ctrl` + `ç` but I get the _Toggle Line Comment_ action
![comment line](https://cloud.githubusercontent.com/assets/243047/20006756/4f4d167e-a299-11e6-9ac4-be88ece244e0.png)

Do you know this behaviour?

**Comment by dayer3** [comment]
I've updated the title because I've just tried this behaviour in Windows 10 x64 and the Command palette shortcuts match with the actions. So in Windows these shortcuts are well.

- Toggle Line Comment
![comment](https://cloud.githubusercontent.com/assets/243047/20031758/f09b4c1a-a37c-11e6-8ee2-41630bb4cafa.png)

- Toggle Integrated Terminal
![terminal](https://cloud.githubusercontent.com/assets/243047/20031762/f6fbb0a4-a37c-11e6-93e7-6a275741f4e8.png)

- Split Editor
![split](https://cloud.githubusercontent.com/assets/243047/20031764/fa429084-a37c-11e6-9c1f-e5efe351aaaa.png)

**Comment by alexandrudima** [comment]
@dayer3 Thank you for the nice report. 

I have just written a wiki page - https://github.com/Microsoft/vscode/wiki/Troubleshooting-Keybindings - to explain in detail what we do with keybindings and how it all works.

May I ask you to give it a quick read such that we end up talking about the same things.

Let's pick one key on your keyboard layout and try to figure out where it all goes wrong on your Fedora machine. I suggest the <kbd>ç</kbd> key.

![Spanish KB](https://cloud.githubusercontent.com/assets/243047/20006354/6791b778-a297-11e6-9884-f5b2552244ae.png)

### 1. Find out the key code "behind" that key

* Go to https://www.w3.org/2002/09/tests/keys.html in Chromium and press <kbd>ç</kbd>
* You should get a specific `keyCode` in the `keydown` column.

### 2. Find out if there's something weird about Chromium

* Open https://www.w3.org/2002/09/tests/keys.html in another browser and press the same key <kbd>ç</kbd>. It should generate the same `keyCode`. If that's not the case, then this might be a Chromium issue.

### 3. Find out the symbolic name for the keyCode
* Use that number and look it up in this table (e.g. let's say it is 222, then it would be `VKEY_OEM_7`) :

![image](https://cloud.githubusercontent.com/assets/5047891/20061969/3e3e1d90-a502-11e6-8c72-650bcdf2fbc9.png)

### 4. Find out what we read from the OS for that keyCode
You will need node.js and a good (node compatible) C++ toolchain.
```
git clone https://github.com/Microsoft/node-native-keymap.git
cd node-native-keymap
npm install -g node-gyp
node-gyp configure
node-gyp build
npm test
```

Look up the results for the previous symbolic key you've identified. They should match with the keyboard layout. If they don't, then this might be an issue with `node-native-keymap`.

**Comment by dayer3** [comment]
@alexandrudima you're welcome

I've follow your steps with the <kbd>ç</kbd> key.

### 1. Find out the key code "behind" that key
- Chromium: 53.0.2785.143
- `keyCode` in the `keydown` column: 191

![ç in Chromium](https://cloud.githubusercontent.com/assets/243047/20081322/3b1ebfe6-a54f-11e6-924d-6ceabae6caff.png)

### 2. Find out if there's something weird about Chromium
- Firefox: 49.0
- `keyCode` in the `keydown` column: 0

![ç in Firefox](https://cloud.githubusercontent.com/assets/243047/20081349/6679190c-a54f-11e6-9c2f-4ad807434ced.png)

### 3. Find out the symbolic name for the keyCode
- According to Chromium: 191 → `VKEY_OEM_2`
- According to Firefox: 0 → I can't find it [in the table](https://msdn.microsoft.com/en-us/library/windows/desktop/dd375731(v=vs.85).aspx).

### 4. Find out what we read from the OS for that keyCode
For `VKEY_OEM_2`:

>   { key_code: 'VKEY_OEM_2',
>     value: '-',
>     withShift: '_',
>     withAltGr: '̣',
>     withShiftAltGr: '̇' },

The `-` value doesn't match with `ç`

I'm going to try this behavior in MS Windows 10.

**Comment by dayer3** [comment]
My test with the <kbd>ç</kbd> key in MS Windows 10 Home x64 (at the same machine).

In Windows the Command palette agrees with shortcuts behaviour:

![code win](https://cloud.githubusercontent.com/assets/243047/20081912/cdf8833a-a552-11e6-9c27-a4ce60d289c5.png)

### 1. Find out the key code "behind" that key
- Chrome: 54.0.2840.71
- `keyCode` in the `keydown` column: 191

![ç in chrome win](https://cloud.githubusercontent.com/assets/243047/20081829/4e207a50-a552-11e6-896c-dbaa2a211c7b.png)


### 2. Find out if there's something weird about Chromium
- Firefox: 49.0.2
- `keyCode` in the `keydown` column: 0

![ç in firefox win](https://cloud.githubusercontent.com/assets/243047/20081843/5a6ea6b0-a552-11e6-83f5-1c90bb6da420.png)

### 3. Find out the symbolic name for the keyCode
- According to Chromium: 191 → `VKEY_OEM_2`
- According to Firefox: 0 → I can't find it in the table.

### 4. Find out what we read from the OS for that keyCode

For `KEY_OEM_2`:

>   { key_code: 'VKEY_OEM_2',
>     value: 'ç',
>     withShift: 'Ç',
>     withAltGr: '}',
>     withShiftAltGr: '' },

So, in Windows this value `ç` match with the original value.

**Comment by dayer3** [comment]
I've tested the behaviour with the <kbd>ç</kbd> key also in Ubuntu 16.10 and the results are the same.

**Comment by alexandrudima** [comment]
@dayer3 Thanks a lot for following through again :).

Based on the mismatch:

---
For `VKEY_OEM_2`:

>   { key_code: 'VKEY_OEM_2',
>     value: '-',
>     withShift: '_',
>     withAltGr: '̣',
>     withShiftAltGr: '̇' },

The `-` value doesn't match with `ç`

---

This, to me, indicates an issue with [`node-native-keymap`](https://github.com/Microsoft/node-native-keymap) on Fedora.

https://github.com/Microsoft/node-native-keymap/blob/master/src/keyboard_x.cc

I have tested that code on Ubuntu, but I have to admit, not at all on Fedora. We do the following:

* we create a single `XKeyEvent`
* we iterate through all the `keyCodes` we're aware of
* change the event's `keycode` and `state` (modifiers)
* we execute `XLookupString` to get a `keysim`
* we then convert that to a unicode string by using a [bunch of tables taken straight from Chromium sources](https://github.com/Microsoft/node-native-keymap/blob/master/deps/chromium/x/keysym_to_unicode.cc#L50)

Are you perhaps knowledgeable in this? Is this somehow wrong/not supported on Fedora? By the fact that you get a return value, it indicates the X Display is configured with standard US...

**Comment by dayer3** [comment]
Sorry @alexandrudima, I don't know about that matter.

I think this could be related with Spanish and Linux, because I've suffered the behavior in three different computers with Fedora and Ubuntu, and every computer with an Spanish keyboard.
If you need more tests I think I could do them, but I'll need you describe to me the steps.

**Comment by alexandrudima** [comment]
@dayer3 Thank you, what would be helpful is if you could document in here what steps you took to configure your installation with Spanish (i.e. I know some folks install 3rd party packages to get different keyboard layouts) such that, when one of us has time, we can install Fedora and configure it in the same way to try to get a reproducible installation.

Maybe someone else from the community is knowledgeable regarding why the X Display would think it sits on US Standard and not on the Spanish keyboard layout. 

fyi @Tyriar if you ever encountered this...

**Comment by Tyriar** [comment]
> I've tested the behavior with the ç key also in Ubuntu 16.10 and the results are the same.

@dayer3 to clarify, you mean they're the same as on Windows (working) or Fedora (not working)?

**Comment by dayer3** [comment]
@alexandrudima: ok I'll take it into account and if I get time I'll describe the installation.

@Tyriar: sorry. I meant neither Ubuntu nor Fedora work.

**Comment by alexandrudima** [comment]
@Tyriar The issue is somewhere in [`node-native-keymap`](https://github.com/Microsoft/node-native-keymap). In there we use X11 apis to find out what's the keyboard layout such that we can render keybindings in the UI with the current keyboard layout. Somehow, that code doesn't work on Fedora + Spanish. I don't know why in Fedora + Spanish the X11 apis return the values usually seen when the keyboard layout is Standard US.

**Comment by dayer3** [comment]
@alexandrudima in Ubuntu + Spanish keyboard the effect is the same. You don't need install Fedora or Ubuntu. You can try it with the live images from official webs over an USB memory. Is for that I've also tried with a live image, without any additional package.
I think this is related directly between [`node-native-keymap`](https://github.com/Microsoft/node-native-keymap) and X11.

According to X11 this is my keyboard configuration:
```
$ setxkbmap -print -verbose 10
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc105
layout:     es,es
variant:    ,
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+es+es:2+inet(evdev)
geometry:   pc(pc105)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+es+es:2+inet(evdev)"	};
	xkb_geometry  { include "pc(pc105)"	};
};
```

**Comment by dayer3** [comment]
I've just tested OpenSUSE Tumbleweed and the problem is the same than Fedora and Ubuntu.

**Comment by alexandrudima** [comment]
I have validated that we now (i.e. after PR #22894 aka #17521 -- in tomorrow's Insiders build) appear to behave correctly for this keyboard layout (at least on Ubuntu, but I believe we are fine on Fedora too)

![image](https://cloud.githubusercontent.com/assets/5047891/24254005/1a0d4018-0fe2-11e7-9c2e-cf5f433c1799.png)


---
### Analysis of default keybinding for `editor.action.commentLine`
* `editor.action.commentLine` is bound by default to the key code binding `` ctrl+/ ``
* `` / `` can be produced through one scan code combination: `` Shift+[Digit7] ``. i.e.:
![image](https://cloud.githubusercontent.com/assets/5047891/24254259/b55d5058-0fe2-11e7-9bdd-b64c7c859a76.png)
* The key code binding `` ctrl+/ `` is therefore mapped to the scan code binding `` Ctrl+Shift+[Digit7] `` and the `F1` command picker now correctly reflects that:
![image](https://cloud.githubusercontent.com/assets/5047891/24254383/12c5d044-0fe3-11e7-946b-383da0cfa8e0.png)
![image](https://cloud.githubusercontent.com/assets/5047891/24254330/ebc733f2-0fe2-11e7-870e-67d3cff502c1.png)
* That is unfortunate, but correct. We don't ship with keyboard layout specific default keybindings, that is something you can upvote in #1240

---
### Analysis of default keybinding for `workbench.action.terminal.toggleTerminal`
* `workbench.action.terminal.toggleTerminal` is bound by default to the key code binding `` ctrl+` ``
* `` ` `` can be produced through one scan code combination: `` [BracketLeft] ``.i.e.
![image](https://cloud.githubusercontent.com/assets/5047891/24254582/9947f174-0fe3-11e7-9f1b-157deef8d640.png)
* Same as above, the key code binding `` ctrl+` `` is therefore mapped to the scan code binding `` Ctrl+[BracketLeft] `` and the `F1` command picker now correctly reflects that:
![image](https://cloud.githubusercontent.com/assets/5047891/24254718/e6a07978-0fe3-11e7-8124-79ae754846c9.png)
![image](https://cloud.githubusercontent.com/assets/5047891/24254689/d6de2c42-0fe3-11e7-84e6-d675864c29c5.png)

---
### Analysis of default keybinding for `workbench.action.splitEditor`
* `workbench.action.splitEditor` is bound by default to the key code binding `` ctrl+\ ``
* `` ` `` can be produced through three scan code combinations: `` Ctrl+Alt+[Minus] ``, `` Ctrl+Alt+[Backquote] `` and `` Ctrl+Alt+Shift+[Backquote] ``. i.e.:
![image](https://cloud.githubusercontent.com/assets/5047891/24254909/58a7ff78-0fe4-11e7-9df1-973847173b39.png)
* because `Ctrl` must be used to produce the visual character, the key code binding `` ctrl+\ `` is not mappable . i.e. the scan code combo would need to be something like `` Ctrl+ Ctrl+Alt+[Minus] ``, which we don't support.
* therefore, the action is unbound by default. That is unfortunate, but correct.

---
### Analysis of pressing `Ctrl+[Backslash]`

![image](https://cloud.githubusercontent.com/assets/5047891/24254994/96909f48-0fe4-11e7-8bc1-ce90072541e9.png)

We now correctly label it:
![image](https://cloud.githubusercontent.com/assets/5047891/24255023/aac3b202-0fe4-11e7-9bfd-dedd323287b2.png)


And it is possible to bind it to `workbench.action.splitEditor` via the following keybinding rule:

```
{
    "key": "ctrl+[Backslash]",
    "command": "workbench.action.splitEditor"
}
```

---
This makes me believe the fix in PR #22894 aka #17521 resolves this issue. Please double check in tomorrow's insiders build.
